trillian/ctfe/storage/storage.go

```diff
@@ -0,0 +1,17 @@
+package storage
+
+import (
+	"context"
+)
+
+// NonLeafCertificateChainStorage is an interface which allows CTFE binaries to use different storage implementations for non-leaf certificate chains.
+type NonLeafCertificateChainStorage interface {
+	// FindAll returns all key-value pairs of non-leaf certificate chains.
+	FindAll(ctx context.Context) (map[string][]byte, error)
+
+	// FindByHash returns the non-leaf certificate chain associated with the provided hash.
+	FindByHash(ctx context.Context, hash string) ([]byte, error)
```

Comment:
Hash is being defined as a string, but what is the encoding? Consider making this `[]byte`, or being clear that this is the hex/base64/other encoding to a string. Passing it through as a `[]byte` is probably the most honest though.

Reply:
Using `[]byte` for the hash is my original bet. However, the LRU cache does not support `[]byte` as the key due to the missing comparable implementation. I would like to align both storage and cache methods to use the same type for the hash argument. The hash type can be updated to `[]byte` and convert that into string in the LRU cache implementation. How does that sound?

The hash is now calculated in the service layer.
https://github.com/roger2hk/certificate-transparency-go/blob/a1aeb9c78da3fe10c24284597e1949080d0940eb/trillian/ctfe/services.go#L148-L151

Reply:
I'm happy with this being a string if it simplifies the implementation. I just ask that the encoding of the `[]byte` to `string` is defined in the docs. As you can see from my last comment, I assumed this would be a hex or base64 encoding, but this is actually a zero-encoding/straight cast between the two types. There's possibly a better way of describing what `string([]byte{})` is than what I've just said, but I'm not able to think of it at the moment. @AlCutter any suggestions?

Reply:
One more data point. Go's map does not support `[]byte` as the key.

Reply:
Yeah, if you're using this as the key in a map then it can't be `[]byte`. My concern was more that this being called `hash` instead of `key` leaves the encoding of this ambiguous. But I can see that you're writing this interface expecting a single client, and multiple implementations. The reason I'm thinking about things that are making this more complex is that I write interfaces expecting multiple clients and multiple implementations. Thus the other clients would need to know how to go back from a `string` to the `hash`.

In the other thread I've suggested making this interface simply a key-value store instead of bound to certs and hashes. If you're happy with that change, then that eliminates all of the discussion in this thread too :-)

Reply:
Updated to use key as the name and `[]byte` as the type.